### ISSUE TITLE
Add no-duplicate-case documentation note about ignoring side effects

### ISSUE THREAD

**Issue by Dionysusnu** [eslint/eslint#15552]
### Environment

Node version: v16.13.1
npm version: 7.13.0
Local ESLint version: v8.7.0
Global ESLint version: N/A
Operating System: Windows 10 Home

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
{
	"parser": "@typescript-eslint/parser",
	"parserOptions": {
		"jsx": true,
		"useJSXTextNode": true,
		"ecmaVersion": 2018,
		"sourceType": "module",
		"project": "./tsconfig.json"
	},
	"plugins": ["@typescript-eslint", "prettier"],
	"extends": ["eslint:recommended"],
	"rules": {}
}
```
</details>

```js
let x = 0;
switch (1) {
	case x++: {
		break;
	}
	case x++: {
		break;
	}
}
```

### What did you expect to happen?

Since the code has side effects, the cases being checked are not identical/duplicate, so no error should be generated.

### What actually happened?

The code generated a false-positive error on cases that are not duplicate.

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I realise this is a very niche case, I found this while enabling eslint on unit tests for a compiler, hence the code testing special edge case behaviours.

**Comment by mdjermanovic** [comment]
Hi @Dionysusnu, thanks for the issue!

I can reproduce this with the default parser as well: [online demo](https://eslint.org/demo#eyJ0ZXh0IjoibGV0IHggPSAwO1xuc3dpdGNoICgxKSB7XG5cdGNhc2UgeCsrOiB7XG5cdFx0YnJlYWs7XG5cdH1cblx0Y2FzZSB4Kys6IHtcblx0XHRicmVhaztcblx0fVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWVsc2UtaWYiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWltcG9ydC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tbm9ub2N0YWwtZGVjaW1hbC1lc2NhcGUiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVuc2FmZS1vcHRpb25hbC1jaGFpbmluZyI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtYmFja3JlZmVyZW5jZSI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19).

This is indeed not a duplicate, but I think it's a good case for `eslint-disable` comments. We could add "When Not To Use It" section in the docs for this rule, mentioning side effects like in [no-dupe-else-if](https://eslint.org/docs/rules/no-dupe-else-if#when-not-to-use-it) rule docs. @eslint/eslint-tsc thoughts?

**Comment by Dionysusnu** [comment]
> I think it's a good case for `eslint-disable` comments

While trying that, I noticed that it seemed impossible to do that in a single comment, because of the other rule `no-fallthrough`. With the code
```ts
switch (1) {
	case x++:
	case x++: {
		break;
	}
}
```
Adding in the eslint-disable comment above the second case, the no-fallthrough rule would start to trigger, probably because it disallows anything but consecutive lines. I didn't find a way to disable both rules in a single comment.

**Comment by nzakas** [comment]
The spirit of the rule is to highlight potential copy-paste errors, which this looks like it could be, so I think the current behavior is the expected behavior and disable comments are the right way to deal with exceptions.

**Comment by btmills** [comment]
> Have you considered linking this demo in the issue format?

@Dionysusnu good idea! #15557 takes a stab at that.

I agree with the rest of the team that side effects feel out of scope for this rule to handle.

**Comment by mdjermanovic** [comment]
> While trying that, I noticed that it seemed impossible to do that in a single comment, because of the other rule `no-fallthrough`. With the code
> 
> ```ts
> switch (1) {
> 	case x++:
> 	case x++: {
> 		break;
> 	}
> }
> ```
> 
> Adding in the eslint-disable comment above the second case, the no-fallthrough rule would start to trigger, probably because it disallows anything but consecutive lines. I didn't find a way to disable both rules in a single comment.

Can you use `eslint-disable-line` on the line where the duplicate is? That way, `no-fallthrough` won't report an error because the two cases stay or consecutive lines.

```js
/* eslint no-duplicate-case: error */
/* eslint no-fallthrough: error */

switch (1) {
	case x++:    
	case x++: { // eslint-disable-line no-duplicate-case
		break;
	}
}
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWR1cGxpY2F0ZS1jYXNlOiBlcnJvciAqL1xuLyogZXNsaW50IG5vLWZhbGx0aHJvdWdoOiBlcnJvciAqL1xuXG5zd2l0Y2ggKDEpIHtcblx0Y2FzZSB4Kys6ICAgIFxuXHRjYXNlIHgrKzogeyAvLyBlc2xpbnQtZGlzYWJsZS1saW5lIG5vLWR1cGxpY2F0ZS1jYXNlXG5cdFx0YnJlYWs7XG5cdH1cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

**Comment by Dionysusnu** [comment]
That works! I thought I'd tried that already, but I guess not properly. Thanks for the help.

**Comment by mdjermanovic** [comment]
I marked this as an accepted issue to update the docs, and prepared PR https://github.com/eslint/eslint/pull/15563
